README.md
# planetmint-driver

CHANGELOG.rst

.. Copyright Planetmint GmbH and Planetmint contributors
   SPDX-License-Identifier: (Apache-2.0 AND CC-BY-4.0)
   Code is Apache-2.0 and docs are CC-BY-4.0

Changelog
=========

0.18.4 (2023-11-07)
-------------------
Changed
^^^^^
* using the all-in-one image for testing purposes
* replace deprecated pysha3 with bultin hashlib

0.18.3 (2023-14-06)
-------------------
Fixed
^^^^^
* bumped versions of several libraries
* fixed security dependency issues

0.18.2 (2023-17-04)
-------------------
Fixed
^^^^^
* adjusted to planetmint 2.4.0 and above changes

0.18.1 (2023-11-04)
-------------------
Fixed
^^^^^
* Transfer preparation limitations
* removed code smell

0.18.0 (2023-23-03)
-------------------
Removed
^^^^^^^
* removed zenroom signing and reflow test cases


0.17.0 (2023-31-01)
-------------------
Changed
^^^^^^^
* adapted to the changed blocks API

0.16.1 (2023-26-01)
-------------------
Changed
^^^^^^^
* using poetry to build and manage the python package

0.16.0 (2023-26-01)
-------------------
Added
^^^^^
* Support for Compose/Decompose specified in PRP-5

0.14.0 (2022-12-02)
-------------------
Fixed
^^^^^
* reintegrated delegated signing support into the driver


0.13.0 (2022-11-28)
-------------------
Changed
^^^^^^^
* adjusted to changes in planetmint-transactions@0.2.2 that are a result from the zenroom upgrade and package renaming

0.12.0 (2022-10-27)
-------------------
Changed
^^^^^^^
* replaced common module with planetmint-transactions package
* adjusted for asset changes in planetmint-transactions@0.2.0

0.11.0 (2022-09-15)
-------------------
Changed
^^^^^^^
* Breaking: fixed the faulty default placemnt of the "data"-tag into assets. Thereby caused obfuscation got removed.
* Breaking: adjusted to the IPLD definitions of planetmint.

0.6.2 (2022-09-08)
------------------
Changed
^^^^^^^
* In setup.py, changed crypto-conditions dependency to above 0.10.0


0.6.2 (2018-11-03)
------------------
Changed
^^^^^^^
* In setup.py, changed python-rapidjson==0.6.0 to ~=0.6.0,
  and changed requests>=2.11.0 to >=2.20.0

0.6.1 (2018-10-21)
------------------
Fixed
^^^^^
* Fixed the problem with a docs page (Handcrafting Transactions) that wouldn't build.

0.6.0 (2018-10-20)
------------------
Changed
^^^^^^^
* Added support for deterministic keypair generation from a 32-byte seed.
  See pull request #487 by external contributor @excerebrose
* Pinned cryptoconditions==0.8.0 in setup.py

Removed
^^^^^^^
* The send() function was removed. See pull request #483.

Known issues
^^^^^^^^^^^^
* Builds of the Handcrafting Transactions page started failing again,
  in Travis CI and on ReadTheDocs.

0.5.3 (2018-09-12)
------------------
Changed
^^^^^^^
* Fixed a failing unit test
* Pinned cryptoconditions==0.7.2 in setup.py
* Fixed the Handcrafting Transactions page in the docs

0.5.2 (2018-08-31)
-------------------
Added
^^^^^

* Cap exponential backoff depending on timeout value for reasonable waiting time in event of network recovery. `#470 <https://github.com/planetmint/planetmint-driver/pull/470>`
* Update cryptoconditions dependency because of security vulnerability CVE-2018-10903. `#472 <https://github.com/planetmint/planetmint-driver/pull/472>`


0.5.1 (2018-08-23)
---------------------
Added
^^^^^

* Support for Planetmint server v2.0.0.b5.
* added round-robin strategy to connect to nodes of the Planetmint network `BEP 14 <https://github.com/planetmint/BEPs/tree/master/14>`_

0.5.0 (2018-06-14)
---------------------
Added
^^^^^
* Added three new methods to send/post a transaction as discussed `here <https://github.com/planetmint/planetmint/issues/2307>`_:

    * ``send_commit``
    * ``send_async``
    * ``send_sync``

Deprecated
^^^^^^^^^^
* ``send()`` under ``TransactionEndpoint``, and available
  via ``Planetmint.transactions``. Replaced by the above three methods:
  ``send_commit()``, ``send_async()``, and ``send_sync()``.


0.5.0a4 (2018-05-07)
---------------------
* `Removed dependencies from Planetmint Server package <https://github.com/planetmint/planetmint-driver/pull/411>`_.


0.5.0a2 (2018-04-18)
---------------------
* `The default mode for sending a transaction is now commit <https://github.com/planetmint/planetmint-driver/issues/386>`_.
* `The metadata endpoint was added <https://github.com/planetmint/planetmint-driver/issues/347>`_.
* Support for Planetmint server v2.0.0a2.


0.5.0a1 (2018-04-03)
--------------------
There were **many** changes between Planetmint 1.3 and Planetmint 2.0 Alpha, too many to list here. We wrote a series of blog posts to summarize most changes, especially those that affect end users and application developers:

* `Some HTTP API Changes in the Next Release <https://blog.planetmint.com/some-http-api-changes-in-the-next-release-49612a537b0c>`_.
* `Three Transaction Model Changes in the Next Release <https://blog.planetmint.com/three-transaction-model-changes-in-the-next-release-dadbac50094a>`_.


0.4.1 (2017-08-02)
------------------
Fixed
^^^^^
* Handcrafting transactions documentation. `Pull request #312
  <https://github.com/planetmint/planetmint-driver/pull/312>`_.
* Quickstart guide. `Pull request #316
  <https://github.com/planetmint/planetmint-driver/pull/316>`_.

0.4.0 (2017-07-05)
------------------
Added
^^^^^
* Support for Planetmint server (HTTP API) 1.0.0.

0.3.0 (2017-06-23)
------------------
Added
^^^^^
* Support for Planetmint server (HTTP API) 1.0.0rc1.
* Support for crypto-conditions RFC draft version 02.
* Added support for text search endpoint ``/assets?search=``

0.2.0 (2017-02-06)
------------------
Added
^^^^^
* Support for Planetmint server 0.9.
* Methods for ``GET /`` and ``GET /api/v1``

Changed
^^^^^^^
* Node URLs, passed to ``Planetmint()`` MUST not include the api prefix
  ``'/api/v1'``, e.g.:

    * BEFORE: ``http://localhost:9984/api/v1``
    * NOW: ``http://localhost:9984``

0.1.0 (2016-11-29)
------------------
Added
^^^^^

planetmint_driver/driver.py
# Copyright Planetmint GmbH and Planetmint contributors
# SPDX-License-Identifier: (Apache-2.0 AND CC-BY-4.0)
# Code is Apache-2.0 and docs are CC-BY-4.0

from .transport import Transport
from .offchain import prepare_transaction, fulfill_transaction
from .utils import normalize_nodes


class Planetmint:
    """A :class:`~planetmint_driver.Planetmint` driver is able to create, sign,
    and submit transactions to one or more nodes in a Federation.

    If initialized with ``>1`` nodes, the driver will send successive
    requests to different nodes in a round-robin fashion (this will be
    customizable in the future).

    """

    def __init__(self, *nodes, transport_class=Transport, headers=None, timeout=20):
        """Initialize a :class:`~planetmint_driver.Planetmint` driver instance.

        Args:
            *nodes (list of (str or dict)): Planetmint nodes to connect to.
                Currently, the full URL must be given. In the absence of any
                node, the default(``'http://localhost:9984'``) will be used.
                If node is passed as a dict, `endpoint` is a required key;
                `headers` is an optional `dict` of headers.
            transport_class: Optional transport class to use.
                Defaults to :class:`~planetmint_driver.transport.Transport`.
            headers (dict): Optional headers that will be passed with
                each request. To pass headers only on a per-request
                basis, you can pass the headers to the method of choice
                (e.g. :meth:`Planetmint().transactions.send_commit()
                <.TransactionsEndpoint.send_commit>`).
            timeout (int): Optional timeout in seconds that will be passed
                to each request.
        """
        self._nodes = normalize_nodes(*nodes, headers=headers)
        self._transport = transport_class(*self._nodes, timeout=timeout)
        self._transactions = TransactionsEndpoint(self)
        self._outputs = OutputsEndpoint(self)
        self._blocks = BlocksEndpoint(self)
        self._assets = AssetsEndpoint(self)
        self._metadata = MetadataEndpoint(self)
        self.api_prefix = "/api/v1"

    @property
    def nodes(self):
        """:obj:`tuple` of :obj:`str`: URLs of connected nodes."""
        return self._nodes

    @property
    def transport(self):
        """:class:`~planetmint_driver.transport.Transport`: Object
        responsible for forwarding requests to a
        :class:`~planetmint_driver.connection.Connection` instance (node).
        """
        return self._transport

    @property
    def transactions(self):
        """:class:`~planetmint_driver.driver.TransactionsEndpoint`:
        Exposes functionalities of the ``'/transactions'`` endpoint.
        """
        return self._transactions

    @property
    def outputs(self):
        """:class:`~planetmint_driver.driver.OutputsEndpoint`:
        Exposes functionalities of the ``'/outputs'`` endpoint.
        """
        return self._outputs

    @property
    def assets(self):
        """:class:`~planetmint_driver.driver.AssetsEndpoint`:
        Exposes functionalities of the ``'/assets'`` endpoint.
        """
        return self._assets

    @property
    def metadata(self):
        """:class:`~planetmint_driver.driver.MetadataEndpoint`:
        Exposes functionalities of the ``'/metadata'`` endpoint.
        """
        return self._metadata

    @property
    def blocks(self):
        """:class:`~planetmint_driver.driver.BlocksEndpoint`:
        Exposes functionalities of the ``'/blocks'`` endpoint.
        """
        return self._blocks

    def info(self, headers=None):
        """Retrieves information of the node being connected to via the
        root endpoint ``'/'``.

        Args:
            headers (dict): Optional headers to pass to the request.

        Returns:
            dict: Details of the node that this instance is connected
            to. Some information that may be interesting:

                * the server version and
                * an overview of all the endpoints

        Note:
            Currently limited to one node, and will be expanded to
            return information for each node that this instance is
            connected to.

        """
        return self.transport.forward_request(method="GET", path="/", headers=headers)

    def api_info(self, headers=None):
        """Retrieves information provided by the API root endpoint
        ``'/api/v1'``.

        Args:
            headers (dict): Optional headers to pass to the request.

        Returns:
            dict: Details of the HTTP API provided by the Planetmint
            server.

        """
        return self.transport.forward_request(
            method="GET",
            path=self.api_prefix,
            headers=headers,
        )


class NamespacedDriver:
    """Base class for creating endpoints (namespaced objects) that can be added
    under the :class:`~planetmint_driver.driver.Planetmint` driver.
    """

    _PATH = "/"

    def __init__(self, driver):
        """Initializes an instance of
        :class:`~planetmint_driver.driver.NamespacedDriver` with the given
        driver instance.

        Args:
            driver (Planetmint): Instance of
                :class:`~planetmint_driver.driver.Planetmint`.
        """
        self.driver = driver

    @property
    def transport(self):
        return self.driver.transport

    @property
    def api_prefix(self):
        return self.driver.api_prefix

    @property
    def rel_uri(self):
        return self.api_prefix + self._PATH


class TransactionsEndpoint(NamespacedDriver):
    """Exposes functionality of the ``'/transactions/'`` endpoint.

    Attributes:
        path (str): The path of the endpoint.

    """

    _PATH = "/transactions/"

    @staticmethod
    def prepare(*, operation="CREATE", signers=None, recipients=None, assets=None, metadata=None, inputs=None):
        """Prepares a transaction payload, ready to be fulfilled.

        Args:
            operation (str): The operation to perform. Must be ``'CREATE'``
                or ``'TRANSFER'``. Case insensitive. Defaults to ``'CREATE'``.
            signers (:obj:`list` | :obj:`tuple` | :obj:`str`, optional):
                One or more public keys representing the issuer(s) of
                the assets being created. Only applies for ``'CREATE'``
                operations. Defaults to ``None``.
            recipients (:obj:`list` | :obj:`tuple` | :obj:`str`, optional):
                One or more public keys representing the new recipients(s)
                of the assets being created or transferred.
                Defaults to ``None``.
            assets (:obj:`dict`, optional): The assets to be created or
                transferred. MUST be supplied for ``'TRANSFER'`` operations.
                Defaults to ``None``.
            metadata (:obj:`dict`, optional): Metadata associated with the
                transaction. Defaults to ``None``.
            inputs (:obj:`dict` | :obj:`list` | :obj:`tuple`, optional):
                One or more inputs holding the condition(s) that this
                transaction intends to fulfill. Each input is expected to
                be a :obj:`dict`. Only applies to, and MUST be supplied for,
                ``'TRANSFER'`` operations.

        Returns:
            dict: The prepared transaction.

        Raises:
            :class:`~.exceptions.PlanetmintException`: If ``operation`` is
                not ``'CREATE'`` or ``'TRANSFER'``.

        .. important::

            **CREATE operations**

            * ``signers`` MUST be set.
            * ``recipients``, ``assets``, and ``metadata`` MAY be set.
            * If ``assets`` is set, it MUST be in the form of::

                {
                    'data': {
                        ...
                    }
                }

            * The argument ``inputs`` is ignored.
            * If ``recipients`` is not given, or evaluates to
              ``False``, it will be set equal to ``signers``::

                if not recipients:
                    recipients = signers

            **TRANSFER operations**

            * ``recipients``, ``assets``, and ``inputs`` MUST be set.
            * ``assets`` MUST be in the form of::

                {
                    'id': '<Asset ID (i.e. TX ID of its CREATE transaction)>'
                }

            * ``metadata`` MAY be set.
            * The argument ``signers`` is ignored.

        """
        return prepare_transaction(
            operation=operation,
            signers=signers,
            recipients=recipients,
            assets=assets,
            metadata=metadata,
            inputs=inputs,
        )

    @staticmethod
    def fulfill(transaction, private_keys):
        """Fulfills the given transaction.

        Args:
            transaction (dict): The transaction to be fulfilled.
            private_keys (:obj:`str` | :obj:`list` | :obj:`tuple`): One or
                more private keys to be used for fulfilling the
                transaction.

        Returns:
            dict: The fulfilled transaction payload, ready to be sent to a
            Planetmint federation.

        Raises:
            :exc:`~.exceptions.MissingPrivateKeyError`: If a private
                key is missing.

        """
        return fulfill_transaction(transaction, private_keys=private_keys)

    def get(self, *, asset_ids, operation=None, headers=None):
        """Given an assets id, get its list of transactions (and
        optionally filter for only ``'CREATE'`` or ``'TRANSFER'``
        transactions).

        Args:
            asset_id (str): Id of the assets.
            operation (str): The type of operation the transaction
                should be. Either ``'CREATE'`` or ``'TRANSFER'``.
                Defaults to ``None``.
            headers (dict): Optional headers to pass to the request.

        Note:
            Please note that the id of an assets in Planetmint is
            actually the id of the transaction which created the assets.
            In other words, when querying for an assets id with the
            operation set to ``'CREATE'``, only one transaction should
            be expected. This transaction will be the transaction in
            which the assets was created, and the transaction id will be
            equal to the given assets id. Hence, the following calls to
            :meth:`.retrieve` and :meth:`.get` should return the same
            transaction.

                >>> bdb = Planetmint()
                >>> bdb.transactions.retrieve('foo')
                >>> bdb.transactions.get(asset_id='foo', operation='CREATE')

            Since :meth:`.get` returns a list of transactions, it may
            be more efficient to use :meth:`.retrieve` instead, if one
            is only interested in the ``'CREATE'`` operation.

        Returns:
            list: List of transactions.

        """
        return self.transport.forward_request(
            method="GET",
            path=self.rel_uri,
            params={"asset_ids": asset_ids, "operation": operation},
            headers=headers,
        )

    def send_async(self, transaction, headers=None):
        """Submit a transaction to the Federation with the mode `async`.

        Args:
            transaction (dict): the transaction to be sent
                to the Federation node(s).
            headers (dict): Optional headers to pass to the request.

        Returns:
            dict: The transaction sent to the Federation node(s).

        """
        return self.transport.forward_request(
            method="POST",
            path=self.rel_uri,
            json=transaction,
            params={"mode": "async"},
            headers=headers,
        )

    def send_sync(self, transaction, headers=None):
        """Submit a transaction to the Federation with the mode `sync`.

        Args:
            transaction (dict): the transaction to be sent
                to the Federation node(s).
            headers (dict): Optional headers to pass to the request.

        Returns:
            dict: The transaction sent to the Federation node(s).

        """
        return self.transport.forward_request(
            method="POST",
            path=self.rel_uri,
            json=transaction,
            params={"mode": "sync"},
            headers=headers,
        )

    def send_commit(self, transaction, headers=None):
        """Submit a transaction to the Federation with the mode `commit`.

        Args:
            transaction (dict): the transaction to be sent
                to the Federation node(s).
            headers (dict): Optional headers to pass to the request.

        Returns:
            dict: The transaction sent to the Federation node(s).

        """
        return self.transport.forward_request(
            method="POST",
            path=self.rel_uri,
            json=transaction,
            params={"mode": "commit"},
            headers=headers,
        )

    def retrieve(self, txid):
        """Retrieves the transaction with the given id.

        Args:
            txid (str): Id of the transaction to retrieve.
            headers (dict): Optional headers to pass to the request.

        Returns:
            dict: The transaction with the given id.

        """
        comp_uri = self.rel_uri + txid
        return self.transport.forward_request(method="GET", path=comp_uri, headers=None)


class OutputsEndpoint(NamespacedDriver):
    """Exposes functionality of the ``'/outputs'`` endpoint.

    Attributes:
        path (str): The path of the endpoint.

    """

    _PATH = "/outputs/"

    def get(self, public_key, spent=None, headers=None):
        """Get transaction outputs by public key. The public_key parameter
        must be a base58 encoded ed25519 public key associated with
        transaction output ownership.

        Args:
            public_key (str): Public key for which unfulfilled
                conditions are sought.
            spent (bool): Indicate if the result set should include only spent
                or only unspent outputs. If not specified (``None``) the
                result includes all the outputs (both spent and unspent)
                associated with the public key.
            headers (dict): Optional headers to pass to the request.

        Returns:
            :obj:`list` of :obj:`str`: List of unfulfilled conditions.

        Example:
            Given a transaction with `id` ``da1b64a907ba54`` having an
            `ed25519` condition (at index ``0``) with alice's public
            key::

                >>> bdb = Planetmint()
                >>> bdb.outputs.get(alice_pubkey)
                ... ['../transactions/da1b64a907ba54/conditions/0']

        """
        return self.transport.forward_request(
            method="GET",
            path=self.rel_uri,
            params={"public_key": public_key, "spent": spent},
            headers=headers,
        )


class BlocksEndpoint(NamespacedDriver):
    """Exposes functionality of the ``'/blocks'`` endpoint.

    Attributes:
        path (str): The path of the endpoint.

    """

    _PATH = "/blocks/"

    def get(self, *, txid, headers=None):
        """Get the block that contains the given transaction id (``txid``)
           else return ``None``

        Args:
            txid (str): Transaction id.
            headers (dict): Optional headers to pass to the request.

        Returns:
            :obj:`list` of :obj:`int`: List of block heights.

        """
        block_list = self.transport.forward_request(
            method="GET",
            path=self.rel_uri,
            params={"transaction_id": txid},
            headers=headers,
        )
        return block_list if block_list else None

    def retrieve(self, block_height):
        """Retrieves the block with the given ``block_height``.

        Args:
            block_height (str): height of the block to retrieve.
            headers (dict): Optional headers to pass to the request.

        Returns:
            dict: The block with the given ``block_height``.

        """
        comp_uri = self.rel_uri + block_height
        return self.transport.forward_request(method="GET", path=comp_uri, headers=None)


class AssetsEndpoint(NamespacedDriver):
    """Exposes functionality of the ``'/assets'`` endpoint.

    Attributes:
        path (str): The path of the endpoint.

    """

    _PATH = "/assets/"

    def get(self, *, cid, limit=0, headers=None):
        """Retrieves the assets that match a given text search string.

        Args:
            search (str): Text search string.
            limit (int): Limit the number of returned documents. Defaults to
                zero meaning that it returns all the matching assets.
            headers (dict): Optional headers to pass to the request.

        Returns:
            :obj:`list` of :obj:`dict`: List of assets that match the query.

        """
        return self.transport.forward_request(
            method="GET",
            path=self.rel_uri + "/" + cid,
            params={"limit": limit},
            headers=headers,
        )


class MetadataEndpoint(NamespacedDriver):
    """Exposes functionality of the ``'/metadata'`` endpoint.

    Attributes:
        path (str): The path of the endpoint.

    """

    _PATH = "/metadata/"

    def get(self, *, search, limit=0, headers=None):
        """Retrieves the metadata that match a given text search string.

        Args:
            search (str): Text search string.
            limit (int): Limit the number of returned documents. Defaults to
                zero meaning that it returns all the matching metadata.
            headers (dict): Optional headers to pass to the request.

        Returns:
            :obj:`list` of :obj:`dict`: List of metadata that match the query.

        """
        return self.transport.forward_request(
            method="GET",
            path=self.rel_uri,
            params={"search": search, "limit": limit},
            headers=headers,
        )
